### PR TITLE
refactor(source): remove `Default` impl from `SourceContext` and `SourceEnumeratorContext`

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

There's actually no **default value** for `SourceContext` and `SourceEnumeratorContext`. When needed in some testing or temporary use cases, the instances created should be considered as **fake or dummy ones** (their field values just don't matter). By removing `Default` trait impl and adding a method called `dummy`, we can prevent incorrect use of such contexts.

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/base.rs [hunk 1]
```diff
@@ -138,27 +138,29 @@ pub struct SourceCtrlOpts {
     pub rate_limit: Option<u32>,
 }
 
-impl Default for SourceCtrlOpts {
```
Comment (BugenZhao) [inline src/connector/src/source/base.rs:141]: May `impl !Default for SourceCtrlOpts {}` and document the reason.

**Review by xxchan — APPROVED** [review]

File: src/connector/src/source/base.rs (hunk 1 above)
Comment (stdrc) [reply src/connector/src/source/base.rs:141]: Good idea! Fixed.

File: src/compute/tests/integration_tests.rs [hunk 2]
```diff
@@ -173,7 +173,10 @@ async fn test_table_materialize() -> StreamResult<()> {
             Arc::new(StreamingMetrics::unused()),
             barrier_rx,
             system_params_manager.get_params(),
-            SourceCtrlOpts::default(),
+            SourceCtrlOpts {
+                chunk_size: 1024,
+                rate_limit: None,
+            },
```
[the remark this answers is not in the document]
Comment (stdrc) [reply src/compute/tests/integration_tests.rs:179]: Because it does matter here. I don't think we should have `SourceCtrlOpts::dummy()`.

**Comment by stdrc** [comment]
> Also please consider making it a const
> 
> ```
> pub const DUMMY: SourceEnumeratorContext = { ... }
> ```

Looks slightly unnecessary to me?

**Comment by tabVersion** [comment]
> > Also please consider making it a const
> > ```
> > pub const DUMMY: SourceEnumeratorContext = { ... }
> > ```
> 
> Looks slightly unnecessary to me?

Because I think you are always referring to a fixed const, calling a func seems unnecessary. It's ok to leave it.

**Pull request merged**